# Get the MNIST data
Get the data using the keras loading function, then process it so it is ready for our fully connected network (FCN).

# Errors?
Check to see that the **kernel** is correct (Python 6.6/Conda 5.2).


In [2]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

short = False
if short:
    train_images = train_images[:7000,:]
    train_labels = train_labels[:7000]
    test_images = test_images[:3000,:]
    test_labels = test_labels[:3000]
#
print("Train info",train_images.shape, train_labels.shape)
print("Test info",test_images.shape, test_labels.shape)
train_images = train_images.reshape((train_images.shape[0],28*28))
train_images = train_images.astype('float32')/255

test_images = test_images.reshape((test_images.shape[0],28*28))
test_images = test_images.astype('float32')/255
from keras.utils import to_categorical

train_labels_cat = to_categorical(train_labels)
test_labels_cat = to_categorical(test_labels)


Train info (60000, 28, 28) (60000,)
Test info (10000, 28, 28) (10000,)


# Build the FCN and train it
We will build a simple 1-hidden-layer network.   We will use 400 hidden nodes since that was close to optimal based on our earlier studies.   We will then save the network for later use.

In [3]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(400,activation='tanh',input_shape=(28*28,)))
network.add(layers.Dense(10,activation='softmax'))
network.compile(optimizer='adam',loss='categorical_crossentropy',metrics=['accuracy'])
#
history = network.fit(train_images,train_labels_cat,epochs=15,batch_size=128,validation_data=(test_images,test_labels_cat))
network.save('fully_trained_model_fcn.h5')

Train on 60000 samples, validate on 10000 samples
Epoch 1/15
60000/60000 [==============================] - 2s 36us/step - loss: 0.3286 - acc: 0.9046 - val_loss: 0.2181 - val_acc: 0.9359
Epoch 2/15
60000/60000 [==============================] - 2s 33us/step - loss: 0.1785 - acc: 0.9489 - val_loss: 0.1477 - val_acc: 0.9556
Epoch 3/15
60000/60000 [==============================] - 2s 33us/step - loss: 0.1233 - acc: 0.9645 - val_loss: 0.1164 - val_acc: 0.9664
Epoch 4/15
60000/60000 [==============================] - 2s 33us/step - loss: 0.0916 - acc: 0.9741 - val_loss: 0.0957 - val_acc: 0.9695
Epoch 5/15
60000/60000 [==============================] - 2s 33us/step - loss: 0.0702 - acc: 0.9794 - val_loss: 0.0840 - val_acc: 0.9742
Epoch 6/15
60000/60000 [==============================] - 2s 34us/step - loss: 0.0535 - acc: 0.9846 - val_loss: 0.0742 - val_acc: 0.9764
Epoch 7/15
60000/60000 [==============================] - 2s 34us/step - loss: 0.0421 - acc: 0.9879 - val_loss: 0.0713 - val_acc

# Finish
The performance on the standard test set is around 98% which is pretty good!